### PR TITLE
Reset power selector on API failure to prevent state mismatch

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/element-hq/element-web/issues/27024

### CODE REVIEW

File: src/components/views/elements/PowerSelector.tsx [hunk 1]
```diff
@@ -106,27 +106,35 @@ export default class PowerSelector<K extends undefined | string> extends React.C
...
             this.setState({ custom: true });
         } else {
             const powerLevel = parseInt(event.target.value);
-            this.props.onChange(powerLevel, this.props.powerLevelKey);
             this.setState({ selectValue: powerLevel });
+            try {
+                await this.props.onChange(powerLevel, this.props.powerLevelKey);
+            } catch {
+                this.initStateFromProps();
+            }
```
Comment (richvdh) [inline src/components/views/elements/PowerSelector.tsx:120]: won't this blow up if the component is unmounted while the callback happens?

Comment (t3chguy) [reply src/components/views/elements/PowerSelector.tsx:120]: Nope, you'll get a warning about setState called on unmounted component, and even that warning is going away in React 18

Comment (richvdh) [reply src/components/views/elements/PowerSelector.tsx:120]: when I said "blow up" I really meant "add to the console warning spam", but TIL about react 18.

Slightly concerned that we'll be on React 17 for a while yet (React 18 was released 2 years ago?) and in the meantime this stuff does add to the red light fatigue in the console, but I guess it's not worth complicating the code to deal with.

Comment (t3chguy) [reply src/components/views/elements/PowerSelector.tsx:120]: Done.

Comment (richvdh) [reply src/components/views/elements/PowerSelector.tsx]: oh, you fixed it anyway.

File: src/components/views/elements/PowerSelector.tsx [hunk 2]
```diff
@@ -60,6 +60,7 @@ export default class PowerSelector<K extends undefined | string> extends React.C
         maxValue: Infinity,
         usersDefault: 0,
     };
+    public unmounted = false;
```
Comment (richvdh) [inline src/components/views/elements/PowerSelector.tsx:63]: can't this be private?

```suggestion
    private unmounted = false;
```

File: src/components/views/elements/PowerSelector.tsx [hunk 3]
```diff
@@ -106,27 +111,37 @@ export default class PowerSelector<K extends undefined | string> extends React.C
...
             this.setState({ custom: true });
         } else {
             const powerLevel = parseInt(event.target.value);
-            this.props.onChange(powerLevel, this.props.powerLevelKey);
             this.setState({ selectValue: powerLevel });
+            try {
+                await this.props.onChange(powerLevel, this.props.powerLevelKey);
+            } catch {
+                if (this.unmounted) return;
+                this.initStateFromProps();
```
Comment (richvdh) [inline src/components/views/elements/PowerSelector.tsx:125]: comment here would be nice

```suggestion
                // If the request failed, roll back the state of the selector.
                this.initStateFromProps();
```

File: src/components/views/elements/PowerSelector.tsx [hunk 4]
```diff
@@ -106,27 +111,37 @@ export default class PowerSelector<K extends undefined | string> extends React.C
...
         event.preventDefault();
         event.stopPropagation();
 
         if (Number.isFinite(this.state.customValue)) {
-            this.props.onChange(this.state.customValue, this.props.powerLevelKey);
+            try {
+                await this.props.onChange(this.state.customValue, this.props.powerLevelKey);
+            } catch {
+                if (this.unmounted) return;
+                this.initStateFromProps();
```
Comment (richvdh) [inline src/components/views/elements/PowerSelector.tsx:143]:
```suggestion
                // If the request failed, roll back the state of the selector.
                this.initStateFromProps();
```

**Review by richvdh — APPROVED** [review]
lgtm otherwise

File: src/components/views/elements/PowerSelector.tsx (hunk 2 above)
Comment (t3chguy) [reply src/components/views/elements/PowerSelector.tsx:63]: Oh yup

**Pull request merged**